torch/csrc/jit/tensorexpr/external_functions.cpp

```diff
@@ -181,6 +183,66 @@ void nnc_aten_triangular_solve(
   }
 }
 
+#ifdef USE_XNNPACK
+
+void nnc_prepacked_linear_clamp_run(
+    int64_t bufs_num,
+    void** buf_data,
+    int64_t* buf_ranks,
+    int64_t* buf_dims,
+    int8_t* buf_dtypes,
+    int64_t args_num,
+    int64_t* extra_args) {
+  using namespace at::native::xnnpack;
+
+  std::vector<at::Tensor> tensors =
+      constructTensors(bufs_num - 1, buf_data, buf_ranks, buf_dims, buf_dtypes);
+
+  const at::Tensor& x = tensors[1];
+  const auto prepacked =
+      c10::intrusive_ptr<LinearOpContext>::unsafe_reclaim_from_nonowning(
+          reinterpret_cast<LinearOpContext*>(buf_data[2]));
+
+  static auto op =
+      c10::Dispatcher::singleton()
+          .findSchemaOrThrow("prepacked::linear_clamp_run", "")
+          .typed<at::Tensor(
+              const at::Tensor&, const c10::intrusive_ptr<LinearOpContext>&)>();
+  at::Tensor output = op.call(x, prepacked);
+  memcpy(
```

Comment:
Intersting, that memcpy is not costing you perf. But I am wondering if you can just do `buf_data[0] = output.data_ptr()`. Maybe not exactly in that form and something rather does exchange data pointers and frees appropriately.

Reply:
Maybe we can add a "context->run_out(input, buf_data[0])" method to write the result into the output buffer directly (we also need to update the buffer size calculation logic to consider the extra memory padding required by the XNNPack).

> if you can just do `buf_data[0] = output.data_ptr()`. Maybe not exactly in that form and something rather does exchange data pointers and frees appropriately.

We cannot simply update the pointer - as you pointed out, the "frees appropriately" part is not trivial. The "output" returned from the aten is a c++ object, which needs to be destructed properly. Right now we manage the life-cycle of all intermediate buffers outside of the assembly function - with this approach we can avoid handling object allocation/free in the assembly code.

Currently all these intermediate buffers are not constructed as "real" aten tensors - the caller simply does "malloc(size)" and "free(buf)". We don't need to ever construct "real" aten tensors for ops that NNC can generate kernels without calling aten kernels (which should become the majority cases). If we want to construct/destruct these intermediate buffers as real tensors, then we need to change the protocol to track more metadata about the tensors (dtype, shape, etc). It's possible but will take some effort.

Maybe eventually we should handle these inside the assembly code (especially when we add support for dynamic shaped inputs), and we will need to have some caching allocator support to avoid memory management cost.

Reply:
"Maybe we can add a "context->run_out(input, buf_data[0])" method to write the result into the output buffer directly (we also need to update the buffer size calculation logic to consider the extra memory padding required by the XNNPack)."

Do you think we'll make all these calls always of "out variant" type? Then make the allocation part of a memory planning stage? Probably then also be more than simple functions (e.g. in the future may need their own shape computation function)?

Reply:
> Do you think we'll make all these calls always of "out variant" type? Then make the allocation part of a memory planning stage? Probably then also be more than simple functions (e.g. in the future may need their own shape computation function)?

Yes I think so. The shape calculation will be more than simple functions when we support dynamic shaped inputs. I don't know whether we want to manage these allocations outside of the assembly code or inside the assembly code.

To do it outside of the assembly code would require some DSL (or maybe with torchscript wrapper?) to specify the shape calculation function for each intermediate buffer.

To do it inside the assembly code we can either directly generate the shape calculation function, or insert a call to the meta-function of structured kernels: https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md. And I imagine we will need to integrate with the caching allocator to avoid repetitive allocations for continuous inferencing.

Reply:
To be clear, doing it inside the assembly code (i.e. allocating buffer inline + calling out-variant) would be similar as directly calling the functional-variant. It's just that the current combination (allocating buffers outside + calling functional-variant + copy data) doesn't make a lot of sense :)

Reply:
I would imagine that you still want to do output allocations outside of the kernel assembly code, mainly so as to allow memory planning be separate from kernels. In this particular case maybe an out variant of the op can provided and used to avoid copying data.

Note that the advantage of using memory allocations outside of the assembly is: 1) we can do liveness analysis and reuse memory as needed. 2) A common memory planner shared between compiler runtime and pt core runtime might be usable.
Main trouble I see with this approach is when allocations are input dependent, such as in maskrcnn where rois are variable.